File: archspot-angular/src/app/core/services/report.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

export interface ReportResponseDTO<T> {
  reportType: string;
  rows: T[];
}

@Injectable({
  providedIn: 'root'
})
export class ReportService {

  private readonly baseUrl = `http://localhost:8080/reports`;

  constructor(private http: HttpClient) { }

  generateReport<T>(filters: any): Observable<ReportResponseDTO<T>> {
    return this.http.post<ReportResponseDTO<T>>(`${this.baseUrl}/generate`, filters);
  }
}

File: ArchSpot-BackEnd/src/main/java/com/archspot/ArchSpot_BackEnd/reports/services/ReportService.java
package com.archspot.ArchSpot_BackEnd.reports.services;

import com.archspot.ArchSpot_BackEnd.reports.dtos.ReportFilterDTO;
import com.archspot.ArchSpot_BackEnd.reports.dtos.ReportResponseDTO;

public interface ReportService {
  ReportResponseDTO<?> generateReport(ReportFilterDTO filters);
}
